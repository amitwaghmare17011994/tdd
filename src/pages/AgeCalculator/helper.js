export const findAge = (date) => {
    const today = new Date();
    const birthDate = date;

    const yearsDifference = today.getFullYear() - birthDate.getFullYear();

    if (
        today.getMonth() < birthDate.getMonth() ||
        (today.getMonth() === birthDate.getMonth() && today.getDate() < birthDate.getDate())
    ) {
        return yearsDifference - 1;
    }

    return yearsDifference;
}