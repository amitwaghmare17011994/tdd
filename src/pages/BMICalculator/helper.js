export const findBMI = (weight, height) => {
    const res =  ((weight) / (height * height))*10000
    return res
}