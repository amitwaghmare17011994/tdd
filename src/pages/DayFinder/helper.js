import moment from 'moment'

const DAYS = {
    0: 'Sunday',
    1: 'Monday',
    2: 'Tuesday',
    3: 'Wednesday',
    4: 'Thursday',
    5: 'Friday',
    6: 'Saturday'
}


export const findDay = (dateObject) => {
    const date = moment(dateObject, "DD/MM/YYYY")
    const dayOfMonth = date.date()
    const year = date.year().toString()
    const month = date.month() + 1
    let lastTwoDigitOfYear = parseInt(year[year.length - 2] + year[year.length - 1])
    const firstTwoDigitOfYear = parseInt(year[0] + year[1])
    let monthNumber = month - 2
    if (monthNumber <= 0) {
        monthNumber = monthNumber === 0 ? 12 : 11
        lastTwoDigitOfYear--
    }


    const resIndex = dayOfMonth + parseInt(parseInt(13 * monthNumber - 1) / 5) + lastTwoDigitOfYear + parseInt(lastTwoDigitOfYear / 4) + parseInt(firstTwoDigitOfYear / 4) - 2 * firstTwoDigitOfYear
    const offset = resIndex % 7
    return DAYS[offset]

}