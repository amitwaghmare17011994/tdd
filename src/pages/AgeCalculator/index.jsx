import { Button, Paper, Typography } from '@material-ui/core'
import React, { useState } from 'react'
import LayoutContainer from '../../components/LayoutContainer'
import { findAge } from './helper'

function AgeCalculator(props) {
    const { callBack = () => { } } = props
    const [selectedDate, setSelectedDate] = useState(null)
    const [res, setRes] = useState('')
    const [error, setError] = useState('')

    const onDateChangeHandler = (e) => {
        const selectedDate = new Date(e.target.value)
        const currentDate = new Date()
        if (selectedDate > currentDate) {
            setError('Enter Past Date')
            return
        }
        setSelectedDate(selectedDate)
        setError('')

    }

    const OnCalculateHandler = () => {
        callBack()
        const age = findAge(selectedDate)
        setRes(age)
    }

    return (
        <LayoutContainer>
            <Paper elevation={3} style={{ height: 500, width: 550, display: 'flex', flexDirection: 'column', justifyContent: 'space-evenly', alignItems: 'center' }}>
                <input onChange={onDateChangeHandler} type={'date'} placeholder={'Enter you birthdate'} />
                 <Button disabled={!!error} onClick={OnCalculateHandler} variant={'contained'} color={'primary'}>Calculate</Button>
                <span id='res'>{res}</span>
                <Typography id='error'>{error}</Typography>
            </Paper>
        </LayoutContainer>
    )
}

export default AgeCalculator
