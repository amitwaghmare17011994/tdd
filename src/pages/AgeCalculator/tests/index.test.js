import { Button } from '@material-ui/core';
import { create, act } from 'react-test-renderer';
import AgeCalculator from '..';
import { findAge } from '../helper';


describe('Age Calculator', () => {
    let root;
    let instance;

    beforeEach(() => {
        root = create(<AgeCalculator />)
        instance = root.root
    })



    it('should render AgeCalculator', () => {
        expect(root.toJSON()).toMatchSnapshot();
    })

    it('should not display anything if date is not selected', () => {
        const button = instance.findByType(Button)
        act(() => {
            button.props.onClick()
        })
        expect(instance.findByProps({ id: 'res' }).props.children).toBe('')

    })

    it('should not select future date', () => {
        const dateInput = instance.findByProps({ type: 'date' })
        act(() => {
            const eventObj = { target: { value: '2021-05-05' } }
            dateInput.props.onChange(eventObj)
        })
        const errorSpan = instance.findByProps({ id: 'error' })
        expect(errorSpan.props.children).toBe('Enter Past Date')
    })

    it('should disable calculate button if date is not valid', () => {
        const button = instance.findByType(Button)
        const dateInput = instance.findByProps({ type: 'date' })
        act(() => {
            const eventObj = { target: { value: '2021-05-05' } }
            dateInput.props.onChange(eventObj)
        })
        expect(button.props.disabled).toBe(true)
    })

    it('should remove error and enable button if date is from past', () => {

        const button = instance.findByType(Button)
        const dateInput = instance.findByProps({ type: 'date' })

        act(() => {
            const eventObj = { target: { value: '2021-05-05' } }
            dateInput.props.onChange(eventObj)
        })

        expect(button.props.disabled).toBe(true)
        const errorSpan = instance.findByProps({ id: 'error' })
        expect(errorSpan.props.children).toBe('Enter Past Date')


        act(() => {
            const eventObj = { target: { value: '2020-05-05' } }
            dateInput.props.onChange(eventObj)
        })

        expect(button.props.disabled).toBe(false)
        expect(errorSpan.props.children).toBe('')

    })

    it('should return return 26 ', () => {
        const bdate = new Date('1995-01-17')
        const result = findAge(bdate)
        expect(result).toBe(26)
    })

    it('should call callBack', () => {

        const fn = jest.fn()
        const root = create(<AgeCalculator callBack={fn} />)
        const instance = root.root

        const button = instance.findByType(Button)

        act(() => {
            button.props.onClick()
        })

        expect(fn).toBeCalled()

    })

})
