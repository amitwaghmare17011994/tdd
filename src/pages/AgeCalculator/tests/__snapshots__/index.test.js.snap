// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Age Calculator should render AgeCalculator 1`] = `
<div
  className="makeStyles-container-1"
>
  <div
    style={
      Object {
        "display": "flex",
        "justifyContent": "center",
        "padding": 20,
      }
    }
  >
    <div
      className="MuiPaper-root MuiPaper-elevation3 MuiPaper-rounded"
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "height": 500,
          "justifyContent": "space-evenly",
          "width": 550,
        }
      }
    >
      <input
        onChange={[Function]}
        placeholder="Enter you birthdate"
        type="date"
      />
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Calculate
        </span>
      </button>
      <span
        id="res"
      >
        
      </span>
      <span
        id="error"
      >
        
      </span>
    </div>
  </div>
</div>
`;
